main/src/ca/uwaterloo/flix/tools/pkg/github/GitHub.scala

```diff
@@ -71,11 +75,169 @@ object GitHub {
     Ok(releaseJsons.arr.map(parseRelease))
   }
 
+  /**
+    * Publish a new release the given project.
+    */
+  def publishRelease(project: Project, version: SemVer, artifacts: Iterable[Path], apiKey: String): Result[Unit, ReleaseError] = {
+    for (
+      _ <- verifyRelease(project, version, apiKey);
+      id <- createDraftRelease(project, version, apiKey);
+      _ <- Result.traverse(artifacts)(p => uploadAsset(p, project, id, apiKey));
+      _ <- markReleaseReady(project, version, id, apiKey)
+    ) yield Ok(())
+  }
+
+  /**
+    * Verifies that the release does not already exist.
+    */
+  private def verifyRelease(project: Project, version: SemVer, apiKey: String): Result[Unit, ReleaseError] = {
+    val url = releaseVersionUrl(project, version)
+
+    try {
+      val conn = url.openConnection().asInstanceOf[HttpsURLConnection]
+      conn.addRequestProperty("Authorization", "Bearer " + apiKey)
+
+      val code = conn.getResponseCode
+      code match {
+        case x if 200 <= x && x <= 299 => Err(ReleaseError.AlreadyExists(project, version))
+        case _ => Ok(())
+      }
+
+    } catch {
+      case _: IOException => Err(ReleaseError.NetworkError)
+    }
+  }
+
+  /**
+    * Create a new release marked as a draft, meaning that it is not publicly visible.
+    * The release will not contain any assets (apart from the default zips of the source code).
+    *
+    * Returns the ID of the release if successful.
+    */
+  private def createDraftRelease(project: Project, version: SemVer, apiKey: String): Result[String, ReleaseError] = {
+    val content: JValue =
+      ("tag_name" -> s"v$version") ~
+        ("name" -> s"v$version") ~
+        ("generate_release_notes" -> true) ~
+        ("draft" -> true)
+
+    val jsonCompact = compact(render(content))
+
+    val url = releasesUrl(project)
+    val json = try {
+      val conn = url.openConnection().asInstanceOf[HttpsURLConnection]
+      conn.setRequestMethod("POST")
+      conn.addRequestProperty("Authorization", "Bearer " + apiKey)
+      conn.setDoOutput(true)
+
+      // Send request
+      val outStream = conn.getOutputStream
+      outStream.write(jsonCompact.getBytes("utf-8"))
+
+      // Process response errors
+      val code = conn.getResponseCode
+      code match {
+        case x if 200 <= x && x <= 299 =>
```

Comment:
We really accept everything from 200 to 299?

Reply:
Would probably be better just to use the documented ones

Reply:
Lets do it. Is 200 not enough?

Reply:
200 and 201 for the different ones